internal/presentation/presentation.go

```diff
@@ -121,6 +121,19 @@ type Output struct {
 	limit             int
 }
 
+type DiscardedOutput struct {
+	Errors            OutputErrors                   `json:"errors,omitempty"`
+	Result            string                         `json:"result,omitempty"`
+	Partial           *rego.PartialQueries           `json:"partial,omitempty"`
+	Metrics           metrics.Metrics                `json:"metrics,omitempty"`
+	AggregatedMetrics map[string]interface{}         `json:"aggregated_metrics,omitempty"`
+	Explanation       []*topdown.Event               `json:"explanation,omitempty"`
+	Profile           []profiler.ExprStats           `json:"profile,omitempty"`
+	AggregatedProfile []profiler.ExprStatsAggregated `json:"aggregated_profile,omitempty"`
+	Coverage          *cover.Report                  `json:"coverage,omitempty"`
+	limit             int
+}
```

Comment:
I wonder if we can do this without a new output struct?

For example, we could have something like [`WithLimit`](https://github.com/open-policy-agent/opa/blob/35efb95b8f74ea6d609dcde4ae4a6479b5aebace/internal/presentation/presentation.go#L125) but `WithDiscard`.

Using `WithDiscard` would set a private `discard` field in [`Output`](https://github.com/open-policy-agent/opa/blob/35efb95b8f74ea6d609dcde4ae4a6479b5aebace/internal/presentation/presentation.go#L111) to true.


Then we can add a function like:

```
func (o Output) MarshalJSON() ([]byte, error) {
	...
}
```

Which reads the `discard` field and if true, replaces the result with a string. Otherwise, it'd just leave it as is.

What do you think?

Reply:
I have created the function

```
func (o Output) MarshalJSON() ([]byte, error) {
	if o.discard {
		var s map[string]interface{} = map[string]interface{} {
			"metrics": o.Metrics,
			"profile": o.Profile,
		}
		return json.Marshal(s)
	} 
	return json.Marshal(o)
}

func (o Output) WithDiscard() Output {
	o.discard = true
	return o
}
```
But, just one doubt, where would the `WithDiscard` function be called?

Reply:
I was also wondering how ti marshal without assigning the `metrics`, `profile` or `result` only, because that won't show errors when there are errors coming in the `eval` output as only `metrics`, `profile` or `result` are assigned

Reply:
That is a very good suggestion, thanks for the pointer
I have updated the code accordingly

Reply:
Hey, thanks for updating this. I think that I confused things with my previous suggestion. What you have is good, but the issue is that we can't set the result field in the JSON to a string, only null.

I had a play with another way that might work here: https://go.dev/play/p/8PFsyunIJIx. This wouldn't need any changes to the output struct (`WithDiscard` etc).

What this does is marshal the output with the result, then unmarshals it back into an untyped structure, then updates the result field to a string, before marshalling it all back to JSON. It's a bit of a round about way of doing it, but I think it works for our purposes. What do you think?

Reply:
Your suggestion totally makes sense and yeah it is a bit of a round way. So, I have tried something new:
```
type OutputWrapper struct {
	Output
}

func (ow OutputWrapper) MarshalJSON() ([]byte, error) {
	if ow.Output.discard {
		ow.Output.Result = nil
		return json.Marshal(ow.Output)
	}
	return json.Marshal(ow.Output)
}

func (e Output) WithDiscard() Output {
	e.discard = true
	return e
}

func Discard(w io.Writer, x interface{}) error {
	encoder := json.NewEncoder(w)
	encoder.SetIndent("", "  ")
	field, ok := x.(Output)
	fieldNew := field.WithDiscard()
	if ok {
		return encoder.Encode(OutputWrapper{Output: fieldNew})
	}
	return encoder.Encode(x)
}
```
This also solves the json marshalling stackoverflow issue
What do you think?

Reply:
`ow.Output.Result = nil` I think the issue is here. Using this approach, we can't set `Result` to a string. Only nil.

Reply:
So, do we want to set it to nil or string?

Reply:
In the issue thread here: https://github.com/open-policy-agent/opa/issues/5863#issuecomment-1623767705 we talked about using the string. I think I still have a preference to show the result as discarded since it's clearer why it's missing than showing nothing at all.

Reply:
Cool, then I will try to replicate this: https://go.dev/play/p/8PFsyunIJIx in the `presentation` file

Reply:
Thanks, let me know how it goes. I think even with the marshal/unmarshal it'll be more contained and easier to change if we need to.

Reply:
Yeah, I think we're going to need to do this outside of the `MarshalJSON ` function, just in the `Discard(w io.Writer, x interface{}) error` is fine I think.

Reply:
It is coming like this:
<img width="1719" alt="Screenshot 2023-07-10 at 4 48 55 PM" src="https://github.com/open-policy-agent/opa/assets/50806654/b2b0e982-bc74-42f3-97d6-a9d0be96e355">

Reply:
You might want to look into using `MarshalIndent`

Reply:
```
> ./opa_darwin_arm64 eval '1/0' --format=discard                                  
{
  "result": "discarded"
}
```

Reply:
And with errors:
```
> ./opa_darwin_arm64 eval '1/0' --format=discard --show-builtin-errors
{
  "errors": [
    {
      "code": "eval_builtin_error",
      "location": {
        "col": 1,
        "file": "",
        "row": 1
      },
      "message": "div: divide by zero"
    }
  ],
  "result": "discarded"
}
```

Reply:
Ok cool, so based on this suggestion here: https://github.com/open-policy-agent/opa/issues/5863#issuecomment-1623763547

I think we can change our logic slightly so that result is only set to "discarded" if there is a result at all. This way, we'll not add discarded in this case `./opa_darwin_arm64 eval '1/0' --format=discard `.

Reply:
Should it be like this:
<img width="812" alt="Screenshot 2023-07-10 at 5 12 15 PM" src="https://github.com/open-policy-agent/opa/assets/50806654/3ec19436-900d-4142-87c8-0ab2444c4631">

Reply:
Cool, what additional changes are required in terms of testing and all?

Reply:
Thanks for working on this 🙂 feel free to push the code once you've done that so I give you another review.

Reply:
Can you kindly review the test cases?

Reply:
I have tried to add the suggestions given by you in the test case file, please review